go/database/vt/memory/trie/nodes.go

```diff
@@ -0,0 +1,462 @@
+	commit() commit.Commitment
+
+	// -- parallel commit support --
+
+	// collectCommitTasks requests the node to append tasks required to be
+	// performed to update its commitment. The resulting list should list the
+	// tasks in order of dependencies, i.e., when processing the list from start
+	// to end, tasks producing inputs for other tasks should appear before
+	// tasks consuming those inputs.
+	collectCommitTasks(tasks *[]*task)
+}
+
+// ---- Inner nodes ----
+
+// inner is the type of an inner node in the Verkle trie. It contains an array
+// of 256 child nodes, indexed by one byte of the key.
+type inner struct {
+	children [256]node
+
+	// The cached commitment of this inner node. It is only valid if the
+	// no dirtyChildCommitments bit is set.
+	commitment commit.Commitment
+
+	// --- Commitment caching ---
+	dirtyChildCommitments bitMap                 // < which children have been modified
+	oldChildCommitments   [256]commit.Commitment // < previous commitments of children
+}
+
+func (i *inner) get(key Key, depth byte) Value {
+	next := i.children[key[depth]]
+	if next == nil {
+		return Value{}
+	}
+	return next.get(key, depth+1)
+}
+
+func (i *inner) set(key Key, depth byte, value Value) node {
+	pos := key[depth]
+	next := i.children[pos]
+	if !i.dirtyChildCommitments.get(pos) {
+		i.dirtyChildCommitments.set(pos)
+		if next != nil {
+			i.oldChildCommitments[pos] = next.commit()
+		}
+	}
+	if next == nil {
+		next = newLeaf(key)
+	}
+	i.children[pos] = next.set(key, depth+1, value)
+	return i
+}
+
+func (i *inner) commit() commit.Commitment {
+	if !i.dirtyChildCommitments.any() {
+		return i.commitment
+	}
+	zone := tracy.ZoneBegin("inner::commit")
+	defer zone.End()
+
+	delta := [commit.VectorSize]commit.Value{}
+	for j := range i.children {
+		if i.dirtyChildCommitments.get(byte(j)) {
+			new := i.children[j].commit().ToValue()
+			old := i.oldChildCommitments[j].ToValue()
+			delta[j] = *new.Sub(old)
+		}
+	}
+
+	// Update the commitment of this inner node.
+	i.commitment.Add(commit.Commit(delta))
+	i.dirtyChildCommitments.clear()
+	return i.commitment
+}
+
+func (i *inner) collectCommitTasks(tasks *[]*task) {
+	if !i.dirtyChildCommitments.any() {
+		return
+	}
+
+	// Produce one task for every dirty child.
+	directChildTasks := make([]*task, 0, 256)
+	delta := [commit.VectorSize]commit.Commitment{}
+	for j := range i.children {
+		if i.dirtyChildCommitments.get(byte(j)) {
+			lengthBefore := len(*tasks)
+			i.children[j].collectCommitTasks(tasks)
+			childTasks := (*tasks)[lengthBefore:]
+
+			var child *task
+			numDependencies := 0
+			if len(childTasks) > 0 {
+				child = childTasks[len(childTasks)-1]
+				numDependencies = 1
+			}
+
+			task := newTask(
+				func() {
+					old := i.oldChildCommitments[j].ToValue()
+					new := i.children[j].commit().ToValue()
+					poly := [commit.VectorSize]commit.Value{}
+					poly[j] = *new.Sub(old)
+					delta[j] = commit.Commit(poly)
+				},
```

Comment:
I seem to recall that you said you would compute the delta commitment for the parent node as part of each node's commitment update (i.e., as the last step of `aggTask` below). If so, why did you move it to a separate task?